File: ios/Aura/AuraProjectPreviewModule.m
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(AuraProjectPreview, NSObject)

RCT_EXTERN_METHOD(
  generatePreview:(NSString *)projectId
  sourceUri:(NSString *)sourceUri
  filterId:(NSString *)filterId
  filterMatrixPayload:(NSString *)filterMatrixPayload
  filterIntensity:(nonnull NSNumber *)filterIntensity
  timeMs:(nonnull NSNumber *)timeMs
  resolve:(RCTPromiseResolveBlock)resolve
  reject:(RCTPromiseRejectBlock)reject
)

@end

File: ios/Aura/AuraProjectPreviewModule.swift
import Foundation
import React

@objc(AuraProjectPreview)
final class AuraProjectPreview: NSObject {
  @objc static func requiresMainQueueSetup() -> Bool {
    false
  }

  @objc(generatePreview:sourceUri:filterId:filterMatrixPayload:filterIntensity:timeMs:resolve:reject:)
  func generatePreview(
    _ projectId: String,
    sourceUri: String,
    filterId: String,
    filterMatrixPayload: String,
    filterIntensity: NSNumber,
    timeMs: NSNumber,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    Task.detached(priority: .utility) {
      do {
        let renderer = await MainActor.run {
          AuraFilteredVideoView(frame: .zero)
        }
        let previewPath = try await renderer.generateProjectPreview(
          projectId: projectId,
          sourceUri: sourceUri,
          filterId: filterId,
          filterMatrixPayload: filterMatrixPayload,
          filterIntensity: CGFloat(truncating: filterIntensity),
          timeMs: Double(truncating: timeMs)
        )

        resolve(previewPath)
      } catch {
        reject("preview_generation_failed", error.localizedDescription, error)
      }
    }
  }
}
